crates/red_knot_python_semantic/src/types.rs

```diff
@@ -996,6 +1018,48 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Returns true if the type does not contain any gradual forms (as a sub-part).
+    pub(crate) fn is_fully_static(self, db: &'db dyn Db) -> bool {
+        match self {
+            Type::Any | Type::Unknown | Type::Todo(_) => false,
+            Type::Never
+            | Type::FunctionLiteral(..)
+            | Type::ModuleLiteral(..)
+            | Type::ClassLiteral(..)
+            | Type::SubclassOf(_)
+            | Type::Instance(_)
```

Comment:
I think for these three, we also need to check the MRO for an Any/Unknown base? A class that inherits Any/Unknown is not a fully static type.

Reply:
Yes, thanks!

For `Type::Instance(_)`, this is obviously correct.

For `Type::ClassLiteral(_)` it took me some time to understand whether we want that or not. I had a good discussion with @AlexWaygood and I believe I now understand why we want that. This also seems to match mypy's and pyright's interpretation:
```py
from typing import Any

class FullyStatic: ...

#instance1: int = FullyStatic()  # obviously doesn't work
#class1: type[int] = FullyStatic  # same

class Gradual(Any): ...

instance2: int = Gradual()  # okay, `Gradual` is a gradual type
class2: type[int] = Gradual  # also okay
```


I then went on to implement this, which seemed fairly trivial at first. But it turns out that this causes all sorts of problems. One of these problems is that we don't fully understand even the most basic types like `str`. In typeshed, `str` inherits from `Sequence[str]`, which is a complex generic protocol. This currently manifests in the following MRO for `str`, which causes us to believe it would be a non-fully-static type.
```py
tuple[Literal[str], Unknown, Literal[object]]
```
This can be patched by matching on a set of known classes. But that is far from ideal. If someone unknowingly pulls in another type from the standard library (outside the set of known classes) with a similar problem, they would see all sorts of weird behavior without a clear indication of what's wrong.

And then there is also another problem where merely iterating over the MRO of a class here (without drawing any conclusions from it) causes us to eagerly infer some types as `Unknown`.

If someone has a good idea on how to resolve this (right now, without waiting for generics), please let me know. Otherwise, I would suggest I add the explanation above as a TODO comment.

Reply:
Fully agree with this analysis and I think the conclusion is right; we need to defer this handling until we can understand most/all typeshed classes that aren't supposed to inherit Any/Unknown, as not inheriting Any/Unknown. TODO comment (and maybe an issue on the backlog, too) seems the right move.